[WIP] - adding plugin for cyberark aim

jakemcdermott opened the pull request:
##### SUMMARY
Adding a plugin for aim. Currently supports Object Query (secret lookup) with optional client certificates if the authentication for the application supports it.

jakemcdermott committed:
add plugin for cyberark aim

jakemcdermott commented on awx/main/credential_plugins/aim.py:
```diff
@@ -0,0 +1,83 @@
+}
+
+def temp_fifo(data):
+    # Open fifo named pipe in a new thread.
```
Certificate data is stored in the db - we're using a fifo here so we can avoid writing it to disk in order to use it in the http request.

We probably don't want to be opening threads if we can avoid it, so I'm currently seeing if we can use a different transport adapter to accept cert data instead of requiring a file path. 

http://docs.python-requests.org/en/master/user/advanced/#transport-adapters

jakemcdermott replied:
After doing some digging, I think this is a deeper problem than what can be solved with a custom transport adapter. You _can_ create and configure your own ssl context object with this approach, but the ssl context api itself [will still require a file path](https://bugs.python.org/issue16487).

Some basic testing has shown the fifo work, and we've used this approach in other places so the current plan is to use it here.

ryanpetrello commented on awx/main/credential_plugins/aim.py:
```diff
@@ -0,0 +1,83 @@
+aim_inputs = {
+    'fields': [{
+        'id': 'url',
+        'label': _('AIM URL'),
```
Not to be confused with "AOL Instant Messager 😂 "
Maybe CyberArk AIM URL?

jakemcdermott replied:
😂good point - fixed

ryanpetrello commented on awx/main/credential_plugins/aim.py:
```diff
@@ -0,0 +1,83 @@
+    client_cert = kwargs.get('client_cert', None)
+    client_key = kwargs.get('client_key', None)
+
+    request_qs = '?AppId={application_id}&Safe={safe}&object={object}'.format(**kwargs)
```
We might need to encode these values since they're in the URL:

```
>>> from urllib.parse import quote
<function quote at 0x7fe2c52d9598>
>>> quote('foo bar')
'foo%20bar'
```

jakemcdermott replied:
👍 I'm sure some or all of these need to be encoded - done.

ryanpetrello commented on awx/main/credential_plugins/aim.py:
```diff
@@ -0,0 +1,83 @@
+    request_qs = '?AppId={application_id}&Safe={safe}&object={object}'.format(**kwargs)
+    request_url = urljoin(url, '/'.join(['AIMWebService', 'api', 'Accounts']))
+
+    client_cert_path = temp_fifo(client_cert)
```
Do we want to check `if client_cert` before doing this?

jakemcdermott replied:
yep

ryanpetrello commented on awx/main/credential_plugins/aim.py:
```diff
@@ -0,0 +1,83 @@
+        client_key_path = temp_fifo(client_key)
+        cert = (client_cert_path, client_key_path)
+
+    res = requests.get(request_url + request_qs, cert=cert, verify=False)
```
`verify=False` probably isn't a secure default to assume (maybe we should make it configurable?)

jakemcdermott replied:
Good idea - I've added this as an input field to the external credential, with default set to `True`.

ryanpetrello reviewed (commented):
This is looking good, a few minor comments.

jakemcdermott committed:
read cert from memory

Certificate data lives in the db. We're using a fifo here so we can avoid writing
it to disk when making the http request.

jakemcdermott committed:
generalize secret textarea hint

jakemcdermott committed:
don't ignore choices param in input config

We always want to use choices if they're available in the input config from the api.

jakemcdermott committed:
add object query fields
